walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,80 @@
 
 package com.google.sps;
 
+import java.util.Arrays;
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    List<TimeRange> requiredEvents = findAttendeeEvents(events, request.getAttendees());
+    List<TimeRange> optionalEvents = findAttendeeEvents(events, request.getOptionalAttendees());
+
+    List<TimeRange> allAttendees = findOpenTimes(requiredEvents, optionalEvents, request);
+    List<TimeRange> onlyOptional = findOpenTimes(Arrays.asList(), optionalEvents, request);
+    List<TimeRange> onlyMandatory = findOpenTimes(requiredEvents, Arrays.asList(), request);
+
+    if (allAttendees.size() > 0) {
+        return allAttendees;
+    } else if (requiredEvents.size() > 0) {
+        return onlyMandatory;
+    } else {
+        return onlyOptional;
+    }
+  }
+
+  private List<TimeRange> findOpenTimes(List<TimeRange> mandatoryAttendeeMeetings, 
+                            List<TimeRange> optionalAttendeeMeetings, MeetingRequest request) {
+
+    ArrayList<TimeRange> options = new ArrayList<TimeRange>();
+    int earliestAvailable = TimeRange.START_OF_DAY;
+
+    List<TimeRange> allMeetings = new ArrayList<TimeRange>(mandatoryAttendeeMeetings);
+    if (optionalAttendeeMeetings != null) {
+        allMeetings.addAll(optionalAttendeeMeetings);
+    }
+
+    Collections.sort(allMeetings, TimeRange.ORDER_BY_START);
+    
+    for(TimeRange eventTime : allMeetings) {
+        if (eventTime.start() >= earliestAvailable) {
+        
+            TimeRange possibleRange = TimeRange.fromStartDuration(earliestAvailable, eventTime.start() - earliestAvailable);
+
+            if (possibleRange.duration() >= request.getDuration()){ 
+                options.add(possibleRange);
+            }
+
+            earliestAvailable = eventTime.end();
+
+        } else if (eventTime.end() > earliestAvailable) {
+            earliestAvailable = eventTime.end();
+        }
+    }
+
+    // Add time after events, if it's able to fit a meeting
+    if (TimeRange.END_OF_DAY - earliestAvailable >= request.getDuration()) {
+        options.add(TimeRange.fromStartEnd(earliestAvailable, TimeRange.END_OF_DAY, true));
+    }    
+
+    return options;
+  }
+
+  private List<TimeRange> findAttendeeEvents(Collection<Event> events, Collection<String> attendeeList) {
+      List<TimeRange> meetingTimes = new ArrayList<TimeRange>();
+
+      for (Event event : events) {
+        for (String person : attendeeList) {
+            if (event.getAttendees().contains(person)) {
```

Comment:
Can possibly add a comment:
Keep the event as long as at least one person in the attendeeList is in it.

Reply:
Suggestion: If you can import Sets from Guava, you can possibly do this and no inside for loop will be needed:
https://www.geeksforgeeks.org/sets-intersection-function-guava-java/
The current way is perfectly good :)